src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocumentIdentity.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Text;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// A small struct that holds the values that define the identity of a source generated document, and don't change
+    /// as new generations happen. This is mostly for convenience as we are reguarly working with this combination of values.
+    /// </summary>
+    internal readonly struct SourceGeneratedDocumentIdentity
+    {
+        public SourceGeneratedDocumentIdentity(DocumentId documentId, string hintName, ISourceGenerator generator, string filePath)
+        {
+            DocumentId = documentId;
+            HintName = hintName;
+            Generator = generator;
+            FilePath = filePath;
+        }
+
+        public DocumentId DocumentId { get; }
+        public string HintName { get; }
+        public ISourceGenerator Generator { get; }
+        public string FilePath { get; }
+
+        public static SourceGeneratedDocumentIdentity Generate(ProjectId projectId, string hintName, ISourceGenerator generator, string filePath)
+        {
+            // We want the DocumentId generated for a generated output to be stable between Compilations; this is so features that track
+            // a document by DocumentId can find it after some change has happened that requires generators to run again.
```

Comment:
is this also stable across vs sesssions?  we def store information in our persistence stores associated with doc-ids.  though i don't think we use their guids.  only their filepaths, and projects.

Reply:
So the GUID portion of a DocumentId or ProjectId isn't stable across VS sessions, so this won't be stable either because it implicitly depends on the ProjectId's GUID.

Reply:
this feels problematic.  every time you start VS we'll create new IDs.  this will definitely bloat up things like our persistence service as it will always seem like there are new files being added.

Reply:
@CyrusNajmabadi I believe the persistence service doesn't use the ProjectId anywhere -- it's always using file names for projects for this reason.